danish/factory.py

```diff
@@ -303,6 +341,215 @@ def _focal_to_pupil(
+
+    Returns
+    -------
+    r_rot : array of float
+        Rotated 3D vector.
+    """
+    sth, cth = np.sin(np.deg2rad(angle)), np.cos(np.deg2rad(angle))
+    x = r[0]*cth + r[1]*sth
+    y = -r[0]*sth + r[1]*cth
+    z = r[2]
+    return np.array([x, y, z])
+
+
+@lru_cache
+def __project_spider_vane(
+    r0, v0, width, length, angle, thx, thy
+):
+    r0 = np.array(r0)
+    v0 = np.array(v0)
+    v0 /= np.linalg.norm(v0)
+
+    # Find direction perp to v0 and (0, 0, 1)
+    # This is the direction along which the width is defined.
+    # Expand the cross product with (0, 0, 1) by hand for speed.
+    norm = np.sqrt(v0[0]**2 + v0[1]**2)
+    perp = np.array([v0[1], -v0[0], 0.0]) / norm
+
+    # Compute the spider vane edges in 3D.
+    centerline = np.array([r0 - v0*length/2, r0 + v0*length/2])
+    edge1 = _rotxy(centerline.T-(perp*width/2)[:, None], angle)
+    edge2 = _rotxy(centerline.T+(perp*width/2)[:, None], angle)
+
+    # Now project the edges onto the entrance pupil (defined by z=0).
+    vproj = _gnomonic(thx, thy)
+    t1 = -edge1[2]/vproj[2]
+    t2 = -edge2[2]/vproj[2]
+    proj1 = (edge1 + np.outer(vproj, t1))
+    proj2 = (edge2 + np.outer(vproj, t2))
+
+    # Mean projected xy position.  I.e., ~center of the spider vane
+    p1 = np.mean(proj1[:2], axis=1)
+    p2 = np.mean(proj2[:2], axis=1)
+
+    sth1 = proj1[1, 1] - proj1[1, 0]
+    cth1 = proj1[0, 1] - proj1[0, 0]
+    norm1 = np.sqrt(sth1**2 + cth1**2)
+    sth1 /= norm1
+    cth1 /= norm1
+
+    sth2 = proj2[1, 1] - proj2[1, 0]
+    cth2 = proj2[0, 1] - proj2[0, 0]
+    norm2 = np.sqrt(sth2**2 + cth2**2)
+    sth2 /= norm2
+    cth2 /= norm2
+
+    return p1[0], p1[1], sth1, cth1, p2[0], p2[1], sth2, cth2
+
+
+def _project_spider_vane(
+    r0, v0, width, length, angle, thx, thy
+):
+    """Project a 3D spider vane onto the entrance pupil.
+
+    Parameters
+    ----------
+    r0 : array of float
+        3D position of the spider vane center (meters).
+    v0 : array of float
+        3D direction of the spider vane (unitless).
+    width : float
+        Width of the spider vane in meters.
+    length : float
+        Length of the spider vane in meters.
+    angle : float
+        Z-axis rotation angle to apply in degrees.
+    thx, thy : float
+        Gnomonic tangent plane coordinates in radians along which to project the
+        spider vane shadow.
+
+    Returns
+    -------
+    p1x, p1y : float
+        Projected ~center position of the first edge of the spider vane (meters).
+    sth1, cth1 : float
+        Projected sine and cosine of the angle of the first edge of the spider vane.
+    p2x, p2y : float
+        Projected ~center position of the second edge of the spider vane (meters).
+    sth2, cth2 : float
+        Projected sine and cosine of the angle of the second edge of the spider vane.
+    """
+    return __project_spider_vane(
+        tuple(r0), tuple(v0), width, length, angle, thx, thy
+    )
+
+
+def strut_masked_fraction(
```

Comment:
Does this ever get used? It looks like the DonutFactory below just calls the version with the underscore?